File: 3.x.x/crates-utopia/validators/src/text.rs
use crate::{Validator, ValueType};
use serde_json::Value;
use std::fmt::Write;

/// Validate string length (and optional character allow-list).
#[derive(Debug, Clone)]
pub struct Text {
    length: usize,
    min: usize,
    allow_list: Vec<char>,
}

impl Text {
    pub fn new(length: usize) -> Self {
        Self {
            length,
            min: 0,
            allow_list: Vec::new(),
        }
    }

    pub fn with_min(mut self, min: usize) -> Self {
        self.min = min;
        self
    }

    pub fn with_allow_list(mut self, list: impl IntoIterator<Item = char>) -> Self {
        self.allow_list = list.into_iter().collect();
        self
    }
}

impl Validator for Text {
    fn description(&self) -> String {
        let mut message = String::from("Value must be a valid string");
        if self.min == self.length && self.length != 0 {
            let _ = write!(message, " and exactly {} chars", self.length);
        } else {
            if self.min != 0 {
                let _ = write!(message, " and at least {} chars", self.min);
            }
            if self.length != 0 {
                let _ = write!(message, " and no longer than {} chars", self.length);
            }
        }
        if !self.allow_list.is_empty() {
            let chars: String = self
                .allow_list
                .iter()
                .map(|c| c.to_string())
                .collect::<Vec<_>>()
                .join(", ");
            let _ = write!(message, " and only consist of '{chars}' chars");
        }
        message
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(s) = value.as_str() else {
            return false;
        };
        let len = s.chars().count();
        if len < self.min {
            return false;
        }
        if self.length != 0 && len > self.length {
            return false;
        }
        if !self.allow_list.is_empty() {
            for ch in s.chars() {
                if !self.allow_list.contains(&ch) {
                    return false;
                }
            }
        }
        true
    }
}

File: packages/validators/src/Validator/Text.php
<?php

declare(strict_types=1);

namespace Utopia\Validator;

use Utopia\Validator;

/**
 * Text
 *
 * Validate that an variable is a valid text value
 */
class Text extends Validator
{
    public const NUMBERS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

    public const ALPHABET_UPPER = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];

    public const ALPHABET_LOWER = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'];

    /**
     * Text constructor.
     *
     * Validate text with maximum length $length. Use $length = 0 for unlimited length.
     * Optionally, provide allowList characters array $allowList to only allow specific character.
     * Set $requireNonBlank to reject values containing only Unicode separators, whitespace, or formatting characters.
     *
     * @param  string[]  $allowList
     */
    public function __construct(
        protected int $length,
        protected int $min = 1,
        protected array $allowList = [],
        protected bool $requireNonBlank = false,
    ) {}

    /**
     * Get Description
     *
     * Returns validator description
     */
    public function getDescription(): string
    {
        $message = 'Value must be a valid string';

        if ($this->min === $this->length) {
            $message .= ' and exactly ' . $this->length . ' chars';
        } else {
            if ($this->min !== 0) {
                $message .= ' and at least ' . $this->min . ' chars';
            }

            if ($this->length !== 0) {
                $message .= ' and no longer than ' . $this->length . ' chars';
            }
        }

        if ($this->allowList) {
            $message .= ' and only consist of \'' . implode(', ', $this->allowList) . '\' chars';
        }

        if ($this->requireNonBlank) {
            $message .= ' and not be blank';
        }

        return $message;
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }

    /**
     * Is valid
     *
     * Validation will pass when $value is text with valid length.
     */
    public function isValid(mixed $value): bool
    {
        if (!\is_string($value)) {
            return false;
        }

        if ($this->requireNonBlank && preg_match('/[^\p{Z}\p{Cf}\s]/u', $value) !== 1) {
            return false;
        }

        if (mb_strlen($value) < $this->min) {
            return false;
        }

        if (mb_strlen($value) > $this->length && $this->length !== 0) {
            return false;
        }

        if (\count($this->allowList) > 0) {
            foreach (str_split($value) as $char) {
                if (!\in_array($char, $this->allowList)) {
                    return false;
                }
            }
        }

        return true;
    }
}
